File: backend/api/src/main/java/com/zeromail/api/error/ErrorCodes.java
package com.zeromail.api.error;

/**
 * Stable, dotted-hierarchy error code constants emitted by {@code GlobalExceptionHandler} in every
 * Phase 1 error response. The frontend switches on these codes (D-C3) and localizes via the {@code
 * errors.*} dictionary namespace; raw exception messages, {@code ProblemDetail.title}, and {@code
 * ProblemDetail.detail} are NEVER user-facing.
 *
 * <p>Naming convention (CONTEXT.md decision D-C3): hierarchical dotted keys, mirroring the {@code
 * messages/{vi,en}.json} {@code errors.*} namespace one-to-one.
 *
 * <p>Adapted from the JHipster {@code ErrorConstants} reference (kept: dotted-key style; rejected:
 * {@code error.http.{status}} default codes — Zero Mail uses semantic keys).
 */
public final class ErrorCodes {

  public static final String AUTH_UNAUTHORIZED = "error.auth.unauthorized";
  public static final String AUTH_FORBIDDEN = "error.auth.forbidden";
  public static final String AUTH_CURRENT_USER_NOT_FOUND = "error.auth.currentUserNotFound";
  public static final String VALIDATION = "error.validation";
  public static final String DATA_INTEGRITY = "error.dataIntegrity";
  public static final String CONFLICT = "error.conflict";
  public static final String BAD_REQUEST = "error.badRequest";
  public static final String GMAIL_DISCONNECTED = "error.gmail.disconnected";
  public static final String AUTH_CONSENT_DENIED = "error.auth.consent_denied";
  public static final String AUTH_GMAIL_SCOPE_REQUIRED = "error.auth.gmail_scope_required";
  public static final String BILLING_INSUFFICIENT_CREDITS = "error.billing.insufficient";
  public static final String BILLING_LEDGER_INVALID_STATE = "error.billing.ledger.invalidState";
  public static final String BILLING_SEPAY_REFERENCE_INVALID =
      "error.billing.sepay.reference_invalid";
  public static final String BILLING_SEPAY_AUTH_INVALID = "error.billing.sepay.auth_invalid";
  public static final String LLM_SAFETY_VIOLATION = "error.llm.safety_violation";
  public static final String LLM_SANITIZATION_FAILED = "error.llm.sanitization_failed";
  public static final String LLM_BYOK_INVALID = "error.llm.byok.invalid";
  public static final String LLM_BYOK_VALIDATE_FAILED = "error.llm.byok.validate_failed";
  public static final String RULES_COMPILE_INVALID = "error.rules.compile.invalid";
  public static final String RULES_COMPILE_CLARIFICATION_REQUIRED =
      "error.rules.compile.clarification_required";
  public static final String RULES_NOT_FOUND = "error.rules.not_found";
  public static final String RULES_PREVIEW_REQUIRED = "error.rules.preview.required";
  public static final String RULES_PREVIEW_INVALID_SAMPLE_SIZE =
      "error.rules.preview.invalid_sample_size";
  public static final String RULES_REORDER_INVALID = "error.rules.reorder.invalid";
  public static final String RULES_VERSION_MISMATCH = "error.rules.version_mismatch";
  public static final String RULES_UNSAFE_ACTION = "error.rules.unsafe_action";
  public static final String RULES_GMAIL_UNAVAILABLE = "error.rules.gmail.unavailable";
  public static final String TRIAGE_UNDO_EXPIRED = "error.triage.undo.expired";
  public static final String TRIAGE_UNDO_ALREADY_DONE = "error.triage.undo.already_done";
  public static final String TRIAGE_UNDO_UNSUPPORTED_ACTION =
      "error.triage.undo.unsupported_action";
  public static final String TRIAGE_SAFETY_VIOLATION = "error.triage.safety_violation";

  private ErrorCodes() {}
}

File: apps/web/lib/api/error-codes.ts
/**
 * Frontend mirror of backend `ErrorCodes.java` constants.
 *
 * Why this exists (REVIEW IN-05):
 *   Switching on bare string literals (`'error.billing.insufficient'` etc.)
 *   in feature code lets a typo silently fall through to the unknown-error
 *   fallback path with no test failure. Centralize the dotted codes here
 *   so a typo is a TypeScript error and so a contract test can fail CI
 *   when this file drifts from `backend/api/.../ErrorCodes.java`.
 *
 * Contract (CONTEXT.md D-C3):
 *   - Backend emits these dotted codes from `GlobalExceptionHandler` /
 *     `ErrorCodes.java` in every error `ApiError.code`.
 *   - Frontend switches on `code` ONLY (never `title` / `detail` / class name).
 *   - Each code maps 1:1 to an `errors.*` translation key after the
 *     leading `error.` prefix is stripped (see `lib/api/errors.ts`).
 *
 * To add a new code:
 *   1. Add the constant to `backend/api/src/main/java/com/zeromail/api/error/ErrorCodes.java`.
 *   2. Add the matching property here.
 *   3. Add the matching translation key under `errors.*` in `i18n/messages/{vi,en}.json`
 *      (typically owned by the feature's `messages.ts`).
 *   4. The contract test in `apps/web/__tests__/error-codes-parity.test.ts`
 *      will fail until both sides agree.
 */

export const ErrorCode = {
  // auth
  AuthUnauthorized: 'error.auth.unauthorized',
  AuthForbidden: 'error.auth.forbidden',
  AuthCurrentUserNotFound: 'error.auth.currentUserNotFound',
  AuthConsentDenied: 'error.auth.consent_denied',
  AuthGmailScopeRequired: 'error.auth.gmail_scope_required',
  // generic
  Validation: 'error.validation',
  DataIntegrity: 'error.dataIntegrity',
  Conflict: 'error.conflict',
  BadRequest: 'error.badRequest',
  // gmail
  GmailDisconnected: 'error.gmail.disconnected',
  // billing
  BillingInsufficient: 'error.billing.insufficient',
  BillingLedgerInvalidState: 'error.billing.ledger.invalidState',
  BillingSepayReferenceInvalid: 'error.billing.sepay.reference_invalid',
  BillingSepayAuthInvalid: 'error.billing.sepay.auth_invalid',
  // llm
  LlmSafetyViolation: 'error.llm.safety_violation',
  LlmSanitizationFailed: 'error.llm.sanitization_failed',
  LlmByokInvalid: 'error.llm.byok.invalid',
  LlmByokValidateFailed: 'error.llm.byok.validate_failed',
  // draft/thread
  DraftGenerationInFlight: 'error.draft.generation.in_flight',
  DraftGenerationUnavailable: 'error.draft.generation.unavailable',
  DraftGenerationFailed: 'error.draft.generation.failed',
  InvalidCursor: 'error.pagination.invalid_cursor',
  // rules
  RulesCompileInvalid: 'error.rules.compile.invalid',
  RulesCompileClarificationRequired: 'error.rules.compile.clarification_required',
  RulesNotFound: 'error.rules.not_found',
  RulesPreviewRequired: 'error.rules.preview.required',
  RulesPreviewInvalidSampleSize: 'error.rules.preview.invalid_sample_size',
  RulesReorderInvalid: 'error.rules.reorder.invalid',
  RulesVersionMismatch: 'error.rules.version_mismatch',
  RulesUnsafeAction: 'error.rules.unsafe_action',
  RulesDuplicate: 'error.rules.duplicate',
  RulesGmailUnavailable: 'error.rules.gmail.unavailable',
} as const;

export type ErrorCodeValue = (typeof ErrorCode)[keyof typeof ErrorCode];
